content/blog/2021/11/deep-dive-into-policy-logging.md

```diff
@@ -0,0 +1,221 @@
+
+```swift
+import Logging
+
+public func validate(payload: String) -> String {
+    // ...
+
+    logger.info("validating object",
+        metadata: [
+            "some-key": "some-value",
+        ])
+
+    // ...
+}
+```
+
+Following the same strategy as the Go and Rust SDK's, the Swift SDK is able to [push log events to
+the host through
+waPC](https://github.com/kubewarden/policy-sdk-swift/blob/59cc979fa9994f1653d3d3e3dd3072188b2c0a18/Sources/kubewardenSdk/logger.swift#L75-L80).
+
+## Cluster Administrators
+
+Being able to log from within a policy is half of the story. Then, we have to be able to read and
+potentially collect these logs.
+
+As we have seen, Kubewarden policies support structured logging that is then forwarded to the
+component running the policy. Usually, this is `kwctl` if you are executing the policy in a manual
+fashion, or the `policy-server` if the policy is being ran in a Kubernetes environment.
+
+Both `kwctl` and the  `policy-server` use the `tracing` crate to produce log events, either the
+events that are produced by the application itself, or by policies that are running in WebAssembly
+runtime environments.
+
+### `kwctl`
+
+The `kwctl` CLI tool takes a very straightforward approach to logging from policies: it will print
+them to the standard error file descriptor.
+
+### `policy-server`
+
+The `policy-server` supports [different log
+formats](https://github.com/kubewarden/policy-server/blob/d615bfe7fdf1fe6001e655360fe85f0db2194410/src/cli.rs#L174-L206):
+`json`, `text` and `otlp`.
+
+`otlp`? I hear you ask. It stands for OpenTelemetry Protocol. We will look into that in a bit.
+
+If the `policy-server` is run with the `--log-fmt` argument set to `json` or `text`, the output will
+be printed to the standard error file descriptor in JSON or plain text formats.
+
+If `--log-fmt` is set to `otlp`, the `policy-server` will use OpenTelemetry to report logs and
+traces.
+
+## OpenTelemetry
+
+Kubewarden is instrumented with OpenTelemetry, so it's possible for the `policy-server` to send
+trace events to an [OpenTelemetry collector](https://opentelemetry.io/docs/collector/) by using the
+OpenTelemtry Protocol (`otlp`).
+
+Our official [Kubewarden Helm
+Chart](https://github.com/kubewarden/helm-charts/blob/199b18b74ca664ab9370ba5fc101f890a3f3d00f/charts/kubewarden-controller/values.yaml)
+has certain values that allow you to deploy Kubewarden with OpenTelemetry support, reporting logs
+and traces to, for example, a Jaeger instance:
+
+```yaml
+telemetry:
+  enabled: True
+  tracing:
+    jaeger:
+      endpoint: "all-in-one-collector.jaeger.svc.cluster.local:14250"
+```
+
+This functionality closes the gap on logging/tracing, given the freedom that the OpenTelemetry
+collector provides to us in terms of flexibility of what to do with this logs and traces.
+
+But this is a big enough topic on its own worth a future blog post. Stay logged!
```

Comment:
Before closing, I would add a link to the section of our docs where we explain how to setup tracing.